File: demo/web_site/src/types/api/signal.ts
/**
 * 信号相关API类型定义
 * 根据 api.json 的 schema 定义
 */

// ==================== 信号类型 ====================

/**
 * 信号类型
 */
export type SignalType = 'buy' | 'sell' | 'close'

/**
 * 信号状态
 */
export type SignalStatus = 'pending' | 'executed' | 'cancelled'

/**
 * 通知类型
 */
export type NotifyType = 'realtime' | 'daily' | 'weekly'

/**
 * 信号记录
 */
export interface Signal {
  id: string
  strategy_id: string
  strategy_name: string | null
  symbol: string
  exchange: string
  signal_type: SignalType
  price: number | null
  quantity: number | null
  confidence: number | null
  timeframe: string | null
  reason: string | null
  indicators: Record<string, any> | null
  status: SignalStatus
  executed_order_id: string | null
  executed_price: number | null
  executed_at: string | null
  is_public: boolean
  subscriber_count: number
  created_at: string
  updated_at: string
}

/**
 * 信号订阅
 */
export interface SignalSubscription {
  id: string
  strategy_id: string
  notify_type: NotifyType
  is_active: boolean
  created_at: string | null
}

/**
 * 订阅请求
 */
export interface SubscribeSignalRequest {
  strategy_id: string
  notify_type?: NotifyType
}

/**
 * 创建信号请求
 */
export interface CreateSignalRequest {
  strategy_id: string
  strategy_name?: string
  symbol: string
  exchange?: string
  signal_type: SignalType
  price: number
  quantity?: number
  confidence?: number
  timeframe?: string
  reason?: string
  indicators?: Record<string, any>
  is_public?: boolean
}

/**
 * 审核信号请求
 */
export interface ApproveSignalRequest {
  is_public: boolean
  reason?: string
}

// ==================== 信号广场相关类型 ====================

/** 信号列表查询参数 */
export interface SignalListParams {
  platform?: string
  type?: 'live' | 'simulated'
  min_days?: number
  search?: string
  sort_by?: 'return_desc' | 'return_asc' | 'drawdown_asc' | 'followers'
  page?: number
  page_size?: number
}

/** 信号列表响应 */
export interface SignalListResponse {
  items: Signal[]
  total: number
  page: number
  pages: number
}

// ==================== 信号详情相关类型 ====================

/** 风险参数 */
export interface RiskParameters {
  maxPositionSize: number
  stopLossPercentage: number
  takeProfitPercentage: number
  riskRewardRatio: number
  volatilityFilter: boolean
}

/** 绩效指标 */
export interface PerformanceMetrics {
  sharpeRatio: number
  winRate: number
  profitFactor: number
  averageHoldingPeriod: number
  maxConsecutiveLosses: number
}

/** 通知设置 */
export interface NotificationSettings {
  emailAlerts: boolean
  pushNotifications: boolean
  telegramBot: boolean
  discordWebhook: boolean
  alertThreshold: number
}

/** 持仓信息 */
export interface SignalPosition {
  symbol: string
  side: 'long' | 'short'
  amount: number
  entryPrice: number
  currentPrice: number
  pnl: number
  pnlPercent: number
}

/** 信号提供者信息 */
export interface SignalProvider {
  id: string
  name: string
  avatar: string
  verified: boolean
  bio: string
  totalSignals: number
  avgReturn: number
  totalFollowers: number
  rating: number
  joinDate: string
  experience: string
  badges: string[]
}

/** 信号详情响应 */
export interface SignalDetailResponse {
  id: string
  name: string
  description: string
  platform: string
  type: 'live' | 'simulated'
  status: 'running' | 'paused' | 'stopped'
  exchange: string
  tradingPair: string
  timeframe: string
  signalFrequency: 'high' | 'medium' | 'low'
  followers: number
  cumulativeReturn: number
  maxDrawdown: number
  runDays: number
  returnCurve: number[]
  returnCurveLabels: string[]
  riskParameters: RiskParameters
  performanceMetrics: PerformanceMetrics
  notificationSettings: NotificationSettings
  positions: SignalPosition[]
  provider: SignalProvider
  createdAt: string
  updatedAt: string
}

/** 收益曲线数据点 */
export interface ReturnCurvePoint {
  date: string
  cumulative_return: number
  drawdown: number
}

/** 收益曲线响应 */
export interface SignalReturnCurveResponse {
  signal_id: string
  period: string
  curve: ReturnCurvePoint[]
}

/** 信号历史记录项 */
export interface SignalHistoryRecord {
  id: string
  action: 'buy' | 'sell'
  symbol: string
  price: number
  amount: number
  time: string
  strength: 'strong' | 'medium' | 'weak'
  pnl: number | null
  status: 'open' | 'closed'
}

/** 信号历史记录响应 */
export interface SignalHistoryResponse {
  total: number
  page: number
  pageSize: number
  items: SignalHistoryRecord[]
}

/** 月度收益项 */
export interface MonthlyReturnItem {
  month: string
  return_rate: number
  trade_count: number
}

/** 月度收益响应 */
export interface SignalMonthlyReturnsResponse {
  signal_id: string
  monthly_returns: MonthlyReturnItem[]
}

/** 回撤曲线数据点 */
export interface DrawdownPoint {
  date: string
  drawdown: number
}

/** 回撤分析响应 */
export interface SignalDrawdownResponse {
  statistics: DrawdownStatistics
  drawdownCurve: number[]
  labels: string[]

}

export interface DrawdownStatistics {
  currentDrawdown: number
  maxDrawdown: number
  avgDrawdown: number
  maxDrawdownDate: string
  maxDrawdownDuration: number
}

/** 用户评价项 */
export interface SignalReview {
  id: string
  user_id: string
  username: string
  avatar?: string
  rating: number
  content: string
  likes: number
  is_liked: boolean
  created_at: string
}

/** 评价列表响应 */
export interface SignalReviewsResponse {
  items: SignalReview[]
  total: number
  page: number
  pages: number
  rating_distribution: Record<string, number>
  average_rating: number
}

/** 提交评价请求 */
export interface SubmitReviewRequest {
  rating: number
  content: string
}

/** 提交评价响应 */
export interface SubmitReviewResponse {
  id: string
  signal_id: string
  rating: number
  content: string
  created_at: string
}

/** 点赞响应 */
export interface ToggleLikeResponse {
  liked: boolean
  total_likes: number
}

/** 创建跟单请求（信号详情页） */
export interface CreateSignalFollowRequest {
  signalId: string
  signalName: string
  exchange: string
  followAmount: number
  followRatio?: number
  stopLoss: number
}

/** 创建跟单响应（信号详情页） */
export interface CreateSignalFollowResponse {
  id: string
  signalId: string
  followAmount: number
  followRatio: number
  stopLoss: number
  status: string
}

// ==================== 跟单列表相关类型 ====================

/** 跟单列表项（对应 GET /api/v1/c/follows/list 响应中的 items） */
export interface FollowListItem {
  id: string
  signalId: string
  signalName: string
  exchange: string
  status: 'following' | 'stopped' | 'paused'
  followAmount: number
  currentValue: number
  totalReturn: number
  todayReturn: number
  maxDrawdown: number
  winRate: number
  followRatio: number
  stopLoss: number
  riskLevel: string
  totalTrades: number
  followDays: number
  startTime: string
  stopTime: string | null
  createTime: string
}

/** 跟单列表响应 */
export interface FollowListResponse {
  records: FollowListItem[]
  total: number
  page: number
  pageSize: number
}

// ==================== 跟单详情相关类型 ====================

/** 跟单持仓 */
export interface FollowPosition {
  id: string
  symbol: string
  side: 'long' | 'short'
  amount: number
  entryPrice: number
  currentPrice: number
  pnl: number
  pnlPercent: number
}

/** 交易点位 */
export interface TradingPoint {
  id: string
  type: 'buy' | 'sell'
  symbol: string
  price: number
  amount: number
  time: string
}

/** 跟单详情响应 */
export interface FollowDetailResponse {
  id: string
  signalId: string
  signalName: string
  tradingPair: string
  exchange: string
  status: 'following' | 'paused' | 'stopped'
  totalReturn: number
  todayReturn: number
  followAmount: number
  currentValue: number
  maxDrawdown: number
  currentDrawdown: number
  followDays: number
  winRate: number
  followRatio: number
  stopLoss: number
  startTime: string
  riskLevel: 'low' | 'medium' | 'high'
  currentPrice: number
  priceChange24h: number
  volume24h: string
  totalTrades: number
  winTrades: number
  lossTrades: number
  avgWin: number
  avgLoss: number
  profitFactor: number
  returnCurve: number[]
  returnCurveLabels: string[]
  positions: FollowPosition[]
  tradingPoints: TradingPoint[]
}

/** 跟单收益对比统计 */
export interface ComparisonStats {
  returnDiff: number
  avgSlippage: number
  copyRate: number
  followFinalReturn: number
  signalFinalReturn: number
}

/** 跟单收益对比响应 */
export interface FollowComparisonResponse {
  follow_id: string
  labels: string[]
  followCurve: number[]
  signalCurve: number[]
  statistics: ComparisonStats
}

/** 跟单交易记录 */
export interface FollowTradeRecord {
  id: string
  side: 'buy' | 'sell'
  symbol: string
  price: number
  amount: number
  total: number
  pnl: number | null
  time: string
  signalTime?: string
  slippage?: number
}

/** 跟单交易记录响应 */
export interface FollowTradesResponse {
  total: number
  page: number
  records: FollowTradeRecord[]
}

/** 跟单事件日志项 */
export interface FollowEvent {
  id: string
  type: 'trade' | 'risk' | 'success' | 'error' | 'system'
  typeLabel: string
  message: string
  time: string
  metadata?: Record<string, any>
}

/** 跟单事件日志响应 */
export interface FollowEventsResponse {
  total: number
  records: FollowEvent[]
}

/** 仓位分布项 */
export interface PositionDistributionItem {
  name: string
  value: number
  percentage?: number
}

/** 仓位详情项 */
export interface PositionDetail {
  symbol: string
  side: 'long' | 'short'
  quantity: number
  entry_price: number
  current_price: number
  unrealized_pnl: number
  ratio: number
}

/** 仓位分布响应 */
export interface FollowPositionsResponse {
  follow_id: string
  positions: PositionDetail[]
  total_value: number
  usage_rate: number
  distribution: PositionDistributionItem[]
}

/** 更新跟单配置请求（信号详情页） */
export interface UpdateSignalFollowConfigRequest {
  followRatio?: number
  stopLoss?: number
  followAmount?: number
}

/** 更新跟单配置响应 */
export interface UpdateSignalFollowConfigResponse {
  follow_id: string
  follow_ratio: number
  stop_loss: number
  follow_amount: number
}

/** 停止跟单请求 */
export interface StopFollowRequest {
  closePositions?: boolean
  reason?: string
}

/** 停止跟单响应 */
export interface StopFollowResponse {
  follow_id: string
  status: string
  closed_positions: boolean
  finalReturn?: number
  stoppedAt?: string
}

// ==================== 策略历史信号类型 ====================

/** 策略历史信号响应 */
export interface StrategySignalHistoryResponse {
  strategy_id: string
  items: Signal[]
  total: number
  page: number
  pages: number
}

/** 订阅响应 */
export interface SignalSubscribeResponse {
  id: string
  strategy_id: string
  notify_type: NotifyType
  is_active: boolean
  message: string
}

/** 订阅列表响应 */
export interface SubscriptionListResponse {
  items: SignalSubscription[]
  total: number
}

/** 审核信号响应 */
export interface ApproveSignalResponse {
  success: boolean
  signal: Signal
  message: string
}

/** 创建信号响应 */
export interface CreateSignalResponse {
  success: boolean
  signal: Signal
  message: string
}

// ==================== K线数据（公共） ====================

/** K线数据点（轻量图表格式） */
export interface KlineDataPoint {
  time: number
  open: number
  high: number
  low: number
  close: number
  volume: number
}

/** K线查询参数（信号页面用） */
export interface SignalKlineParams {
  symbol: string
  interval: string
  limit?: number
  startTime?: number
  endTime?: number
}

/** K线响应（信号页面用） */
export interface SignalKlineResponse {
  timestamp: number
  open: number
  high: number
  low: number
  close: number
  volume: number
}

File: demo/quant_trading_system/src/quant_trading_system/models/signal.py
"""
信号模块数据库模型
=================

包含信号广场相关的所有 ORM 模型：
- Signal: 信号广场主表
- SignalProvider: 信号提供者
- SignalReview / SignalReviewLike: 用户评价
- SignalRiskParameters: 信号风险参数
- SignalPerformanceMetrics: 信号绩效指标
- SignalNotificationSettings: 信号通知设置
- SignalPosition: 信号当前持仓
- SignalTradeRecord: 信号交易记录
- SignalMonthlyReturn: 信号月度收益
- SignalReturnCurve: 信号收益曲线
- SignalSubscription: 用户订阅信号
"""

from datetime import datetime

from sqlalchemy import Column, String, Boolean, DateTime, Text, Integer, \
    BigInteger, \
    Numeric
from sqlalchemy.dialects.postgresql import JSONB
from sqlalchemy.orm import relationship

from quant_trading_system.core.snowflake import generate_snowflake_id
from quant_trading_system.models.base import Base


class Signal(Base):
    """
    信号广场主表模型 — 信号源的核心元数据

    信号来源分为两类：
    - strategy: 交易所跟单信号（由系统策略产生）
    - subscribe: 订阅大佬账户（通过 API 授权监听目标账户仓位变化）
    """
    __tablename__ = "signal"

    id = Column(BigInteger, primary_key=True, default=generate_snowflake_id)
    name = Column(String(128), nullable=False)
    platform = Column(String(64), nullable=False, default="Binance")
    type = Column(String(16), nullable=False, default="live")  # live / simulated
    signal_source = Column(String(16), nullable=False,
                           default="strategy")  # strategy / subscribe
    status = Column(String(16), nullable=False,
                    default="running")  # running / paused / stopped
    exchange = Column(String(64))
    account_type = Column(String(16), default="spot")  # spot / futures
    trading_pair = Column(String(32))
    timeframe = Column(String(16))
    signal_frequency = Column(String(16))  # high / medium / low
    description = Column(Text)
    provider_id = Column(BigInteger)  # 软关联 -> signal_providers.id
    strategy_id = Column(String(128))

    # ── 订阅大佬账户时的 API 授权信息（signal_source='subscribe' 时必填）──
    target_api_key = Column(String(512))
    target_api_secret = Column(String(512))
    target_passphrase = Column(String(512))
    target_account_name = Column(String(256))
    testnet = Column(Boolean, default=False)

    # ── WebSocket 监听配置 ──
    auto_start_stream = Column(Boolean, default=True)
    watch_symbols = Column(JSONB)
    sync_history = Column(Boolean, default=False)

    # ── 统计字段 ──
    followers_count = Column(Integer, nullable=False, default=0)
    run_days = Column(Integer, nullable=False, default=0)
    cumulative_return = Column(Numeric(12, 4), nullable=False, default=0)
    max_drawdown = Column(Numeric(12, 4), nullable=False, default=0)
    created_at = Column(DateTime, nullable=False, default=datetime.utcnow)
    updated_at = Column(DateTime, nullable=False, default=datetime.utcnow)
    enable_flag = Column(Boolean, default=True)

    # 关系
    provider = relationship("SignalProvider", backref="signals",
                            foreign_keys=[provider_id],
                            primaryjoin="Signal.provider_id == SignalProvider.id")
    risk_parameters = relationship("SignalRiskParameters", uselist=False,
                                   back_populates="signal",
                                   cascade="all, delete-orphan",
                                   foreign_keys="[SignalRiskParameters.signal_id]",
                                   primaryjoin="Signal.id == SignalRiskParameters.signal_id")
    performance_metrics = relationship("SignalPerformanceMetrics", uselist=False,
                                       back_populates="signal",
                                       cascade="all, delete-orphan",
                                       foreign_keys="[SignalPerformanceMetrics.signal_id]",
                                       primaryjoin="Signal.id == SignalPerformanceMetrics.signal_id")
    notification_settings = relationship("SignalNotificationSettings", uselist=False,
                                         back_populates="signal",
                                         cascade="all, delete-orphan",
                                         foreign_keys="[SignalNotificationSettings.signal_id]",
                                         primaryjoin="Signal.id == SignalNotificationSettings.signal_id")
    positions = relationship("SignalPosition", back_populates="signal",
                             cascade="all, delete-orphan",
                             foreign_keys="[SignalPosition.signal_id]",
                             primaryjoin="Signal.id == SignalPosition.signal_id")
    trade_records = relationship("SignalTradeRecord", back_populates="signal",
                                 cascade="all, delete-orphan",
                                 foreign_keys="[SignalTradeRecord.signal_id]",
                                 primaryjoin="Signal.id == SignalTradeRecord.signal_id")


class SignalProvider(Base):
    """信号提供者模型"""
    __tablename__ = "signal_providers"

    id = Column(BigInteger, primary_key=True, default=generate_snowflake_id)
    user_id = Column(BigInteger, nullable=False)
    name = Column(String(128), nullable=False)
    avatar = Column(String(512))
    verified = Column(Boolean, default=False)
    bio = Column(Text)
    total_signals = Column(Integer, default=0)
    avg_return = Column(Numeric(10, 6), default=0)
    total_followers = Column(Integer, default=0)
    rating = Column(Numeric(3, 2), default=0)
    experience = Column(String(64))
    badges = Column(JSONB, default=[])
    create_time = Column(DateTime, default=datetime.utcnow)
    update_time = Column(DateTime, default=datetime.utcnow)
    enable_flag = Column(Boolean, default=True)

    # 关系
    user = relationship("User", foreign_keys=[user_id],
                        primaryjoin="SignalProvider.user_id == User.id")


class SignalReview(Base):
    """用户评价模型"""
    __tablename__ = "signal_reviews"

    id = Column(BigInteger, primary_key=True, default=generate_snowflake_id)
    signal_id = Column(BigInteger, nullable=False)
    user_id = Column(BigInteger, nullable=False)
    rating = Column(Integer, nullable=False)
    content = Column(Text, nullable=False)
    likes = Column(Integer, default=0)
    status = Column(String(16), default="active")
    create_time = Column(DateTime, default=datetime.utcnow)
    update_time = Column(DateTime, default=datetime.utcnow)

    # 关系
    user = relationship("User", foreign_keys=[user_id],
                        primaryjoin="SignalReview.user_id == User.id")


class SignalReviewLike(Base):
    """评价点赞模型"""
    __tablename__ = "signal_review_likes"

    id = Column(BigInteger, primary_key=True, default=generate_snowflake_id)
    review_id = Column(BigInteger, nullable=False)
    user_id = Column(BigInteger, nullable=False)
    create_time = Column(DateTime, default=datetime.utcnow)


class SignalRiskParameters(Base):
    """信号风险参数模型"""
    __tablename__ = "signal_risk_parameters"

    id = Column(BigInteger, primary_key=True, default=generate_snowflake_id)
    signal_id = Column(BigInteger, nullable=False, unique=True)
    max_position_size = Column(Numeric(8, 2))
    stop_loss_percentage = Column(Numeric(8, 2))
    take_profit_percentage = Column(Numeric(8, 2))
    risk_reward_ratio = Column(Numeric(8, 2))
    volatility_filter = Column(Boolean, default=False)
    created_at = Column(DateTime, nullable=False, default=datetime.utcnow)
    updated_at = Column(DateTime, nullable=False, default=datetime.utcnow)

    signal = relationship("Signal", back_populates="risk_parameters",
                          foreign_keys=[signal_id],
                          primaryjoin="SignalRiskParameters.signal_id == Signal.id")


class SignalPerformanceMetrics(Base):
    """信号绩效指标模型"""
    __tablename__ = "signal_performance_metrics"

    id = Column(BigInteger, primary_key=True, default=generate_snowflake_id)
    signal_id = Column(BigInteger, nullable=False, unique=True)
    sharpe_ratio = Column(Numeric(8, 4))
    win_rate = Column(Numeric(8, 4))
    profit_factor = Column(Numeric(8, 4))
    average_holding_period = Column(Numeric(8, 2))
    max_consecutive_losses = Column(Integer)
    total_trades = Column(Integer, default=0)
    win_trades = Column(Integer, default=0)
    loss_trades = Column(Integer, default=0)
    avg_win = Column(Numeric(14, 4))
    avg_loss = Column(Numeric(14, 4))
    created_at = Column(DateTime, nullable=False, default=datetime.utcnow)
    updated_at = Column(DateTime, nullable=False, default=datetime.utcnow)

    signal = relationship("Signal", back_populates="performance_metrics",
                          foreign_keys=[signal_id],
                          primaryjoin="SignalPerformanceMetrics.signal_id == Signal.id")


class SignalNotificationSettings(Base):
    """信号通知设置模型"""
    __tablename__ = "signal_notification_settings"

    id = Column(BigInteger, primary_key=True, default=generate_snowflake_id)
    signal_id = Column(BigInteger, nullable=False, unique=True)
    email_alerts = Column(Boolean, default=True)
    push_notifications = Column(Boolean, default=True)
    telegram_bot = Column(Boolean, default=False)
    discord_webhook = Column(Boolean, default=False)
    alert_threshold = Column(Numeric(8, 2))
    created_at = Column(DateTime, nullable=False, default=datetime.utcnow)
    updated_at = Column(DateTime, nullable=False, default=datetime.utcnow)

    signal = relationship("Signal", back_populates="notification_settings",
                          foreign_keys=[signal_id],
                          primaryjoin="SignalNotificationSettings.signal_id == Signal.id")


class SignalPosition(Base):
    """信号当前持仓模型"""
    __tablename__ = "signal_position"

    id = Column(BigInteger, primary_key=True, default=generate_snowflake_id)
    signal_id = Column(BigInteger, nullable=False)
    symbol = Column(String(32), nullable=False)
    side = Column(String(8), nullable=False)
    amount = Column(Numeric(18, 8), nullable=False)
    entry_price = Column(Numeric(18, 8), nullable=False)
    current_price = Column(Numeric(18, 8))
    pnl = Column(Numeric(14, 4))
    pnl_percent = Column(Numeric(10, 4))
    status = Column(String(16), nullable=False, default="open")
    opened_at = Column(DateTime, nullable=False, default=datetime.utcnow)
    closed_at = Column(DateTime)
    updated_at = Column(DateTime, nullable=False, default=datetime.utcnow)

    signal = relationship("Signal", back_populates="positions",
                          foreign_keys=[signal_id],
                          primaryjoin="SignalPosition.signal_id == Signal.id")


class SignalTradeRecord(Base):
    """信号交易记录模型"""
    __tablename__ = "signal_trade_record"

    id = Column(BigInteger, primary_key=True, default=generate_snowflake_id)
    signal_id = Column(BigInteger, nullable=False)
    original_order_id = Column(String(64), nullable=True, index=True)
    order_status = Column(String(24), nullable=True)
    action = Column(String(8), nullable=False)
    symbol = Column(String(32), nullable=False)
    price = Column(Numeric(18, 8), nullable=False)
    amount = Column(Numeric(18, 8), nullable=False)
    total = Column(Numeric(18, 4))
    strength = Column(String(8))
    pnl = Column(Numeric(14, 4))
    open_trade_id = Column(BigInteger, nullable=True)
    traded_at = Column(DateTime, nullable=False, timezone=True)
    created_at = Column(DateTime, nullable=False, timezone=True, default=datetime.utcnow)

    signal = relationship("Signal", back_populates="trade_records",
                          foreign_keys=[signal_id],
                          primaryjoin="SignalTradeRecord.signal_id == Signal.id")
    open_trade = relationship("SignalTradeRecord", remote_side=[id],
                              foreign_keys=[open_trade_id],
                              primaryjoin="SignalTradeRecord.open_trade_id == SignalTradeRecord.id")


class SignalMonthlyReturn(Base):
    """信号月度收益模型"""
    __tablename__ = "signal_monthly_return"

    signal_id = Column(BigInteger, primary_key=True)
    month = Column(DateTime, primary_key=True)
    return_value = Column(Numeric(12, 4), nullable=False)


class SignalReturnCurve(Base):
    """信号收益曲线模型（时序表）"""
    __tablename__ = "signal_return_curve"

    signal_id = Column(BigInteger, primary_key=True)
    time = Column(DateTime, primary_key=True)
    return_value = Column(Numeric(12, 4), nullable=False)
    drawdown = Column(Numeric(12, 4))


class SignalSubscription(Base):
    """用户订阅信号模型"""
    __tablename__ = "signal_subscriptions"

    id = Column(BigInteger, primary_key=True, default=generate_snowflake_id)
    user_id = Column(BigInteger, nullable=False)
    strategy_id = Column(String(128), nullable=False)
    notify_type = Column(String(32), default="realtime")
    is_active = Column(Boolean, default=True)
    created_at = Column(DateTime, default=datetime.utcnow)
    updated_at = Column(DateTime, default=datetime.utcnow)
